flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java

```diff
@@ -222,6 +222,10 @@ MainThreadExecutor getMainThreadExecutorForTesting() {
 			return super.getMainThreadExecutor();
 		}
 
+		void setTestingNMClientAsync(TestingNMClientAsync testingNMClientAsync) {
+			this.testingNMClientAsync = testingNMClientAsync;
+		}
```

Comment:
Currently, I don't see the benefit or requirement of changing the behavior during tests. Besides, some of its functions are called asynchronously. I prefer to add `TestingNMClientAsync` as a parameter of the constructor of `Context`. WDYT?

Reply:
I'm not sure about that. There are quite some fields in `Context`. If for each field that needs to be customized we always add it to the `Context` constructor parameters, then the constructor will soon become complex, or we have to maintain lots of constructors with different parameter combinations.

I think the key point of making `Consumer` and `Function` of `TestingNMClientAsync` is that, we can use the same default `TestingNMClientAsync` created in `Context`, but still be able to define its behavior in the test case. Being able to change it's behavior is only an extra benefit.

Reply:
My concern for defining the behavior in the test case is:
- If someone sets the `startContainerAsync`, which is called asynchronously, multiple times during the test, it may cause an unknown issue which is hard to debug.

Regarding the constructor of Context, I think we could fix it by adding a builder class.

Reply:
Regarding your concern:
- In most cases, people can simply set the functions before starting the RM, i.e., before `runTest`.
- In cases changing functions during the test, one can easily make sure the change of the function happens after calls to the previous function with `CompletableFuture`. Similar to what we do with `rpcService.(re)setRpcGatewayFutureFunction` in `ResourceManagerTaskExecutorTest#testDelayedRegisterTaskExecutor`. 
- Even if someone messed it up, it should only affect certain test case, which is easy to locate. No production code or other test cases will be affected.

I agree that adding a builder class for `Context` could also be an option. However, I think overwriting the `TestingNMClientAsync` functions might be a better solution, for the following advantages compared to adding builder for `Context`:
- It narrows down the scope that we need to customize in a test case. We need only provide a function that defines the NM client behavior (exactly what actually needs to be customized for the test case), instead of the entire `TestingNMClientAsync`.
- It provides the flexibility of changing NM client behavior during the test, which might not be needed ATM but might be in the future.
- Currently we define the test cases in double brace initialization of the `Context` class, providing very good readability. I believe the double brace initialization is required to directly follow the constructor, which means adding a builder to `Context` would prevent the usage of it.

Reply:
Agreed. We can leverage `CompletableFuture`.

Regarding creating `TestingNMClientAsync` in `Context`, we need to create a `TestingYarnResourceManager` first and then pass it as the `callBackHandler` to the constructor of `TestingNMClientAsync`.
So, I prefer to create `TestingNMClientAsync` in the constructor of `TestingYarnResourceManager` and add a "getter" function in `TestingYarnResourceManager`. WDYT?